File: app/src/main/kotlin/ua/kyiv/putivnyk/ui/viewmodel/EventsViewModel.kt
package ua.kyiv.putivnyk.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.work.WorkManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import ua.kyiv.putivnyk.data.model.SyncStatus
import ua.kyiv.putivnyk.data.model.RoutePoint
import ua.kyiv.putivnyk.data.remote.events.EventsBackendRepository
import ua.kyiv.putivnyk.data.remote.events.model.EventItem
import ua.kyiv.putivnyk.data.repository.RouteRepository
import ua.kyiv.putivnyk.data.repository.SyncStateRepository
import ua.kyiv.putivnyk.data.repository.UserPreferenceRepository
import ua.kyiv.putivnyk.domain.usecase.routing.RouteMetricsCalculator
import java.util.Locale
import ua.kyiv.putivnyk.sync.EventsSyncWorker
import ua.kyiv.putivnyk.sync.SyncScheduler
import javax.inject.Inject

@HiltViewModel
class EventsViewModel @Inject constructor(
    private val eventsRepository: EventsBackendRepository,
    private val userPreferenceRepository: UserPreferenceRepository,
    private val routeRepository: RouteRepository,
    private val syncStateRepository: SyncStateRepository,
    private val workManager: WorkManager
) : ViewModel() {

    private val supportedEventLanguages = setOf("uk", "en")

    enum class EventSortMode {
        DATE,
        TITLE,
        CATEGORY
    }

    private val syncEntity = "events_backend"
    private val staleThresholdMs = 24 * 60 * 60 * 1000L

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _selectedCategory = MutableStateFlow<String?>(null)
    val selectedCategory: StateFlow<String?> = _selectedCategory.asStateFlow()

    private val _sortMode = MutableStateFlow(EventSortMode.DATE)
    val sortMode: StateFlow<EventSortMode> = _sortMode.asStateFlow()

    private val _events = MutableStateFlow<List<EventItem>>(emptyList())
    private val _selectedEvent = MutableStateFlow<EventItem?>(null)
    val selectedEvent: StateFlow<EventItem?> = _selectedEvent.asStateFlow()

    val availableCategories: StateFlow<List<String>> = _events
        .map { list ->
            list.map { it.category.trim() }
                .filter { it.isNotBlank() }
                .distinct()
                .sorted()
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = emptyList()
        )

    val syncStatusLabel: StateFlow<String> = syncStateRepository.observeAll()
        .map { states ->
            val current = states.firstOrNull { it.entityName == syncEntity }
            when (current?.status ?: SyncStatus.IDLE) {
                SyncStatus.IDLE -> "events.sync_idle"
                SyncStatus.RUNNING -> "events.sync_running"
                SyncStatus.SUCCESS -> "events.sync_success"
                SyncStatus.ERROR -> "events.sync_error"
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = "events.sync_idle"
        )

    val freshnessLabel: StateFlow<String> = syncStateRepository.observeAll()
        .map { states ->
            val current = states.firstOrNull { it.entityName == syncEntity }
            when {
                current == null || current.lastSyncAt == 0L -> "events.freshness_unknown"
                System.currentTimeMillis() - current.lastSyncAt > staleThresholdMs -> "events.freshness_stale"
                else -> "events.freshness_ok"
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = "events.freshness_unknown"
        )

    val isDataStale: StateFlow<Boolean> = syncStateRepository.observeAll()
        .map { states ->
            val current = states.firstOrNull { it.entityName == syncEntity }
            current == null || current.lastSyncAt == 0L || (System.currentTimeMillis() - current.lastSyncAt > staleThresholdMs)
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = false
        )

    val events: StateFlow<List<EventItem>> = combine(
        _events,
        _searchQuery,
        _selectedCategory,
        _sortMode
    ) { list, query, selectedCategory, sortMode ->
        var filtered = list
        if (query.isNotBlank()) {
            filtered = filtered.filter { item ->
            item.title.contains(query, ignoreCase = true) ||
                item.description.contains(query, ignoreCase = true) ||
                item.category.contains(query, ignoreCase = true) ||
                item.locationName.contains(query, ignoreCase = true)
            }
        }

        if (!selectedCategory.isNullOrBlank()) {
            filtered = filtered.filter { it.category.equals(selectedCategory, ignoreCase = true) }
        }

        when (sortMode) {
            EventSortMode.DATE -> filtered.sortedByDescending { it.startsAt }
            EventSortMode.TITLE -> filtered.sortedBy { it.title.lowercase() }
            EventSortMode.CATEGORY -> filtered.sortedBy { it.category.lowercase() }
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = emptyList()
    )

    init {
        refresh()
    }

    fun updateSearchQuery(value: String) {
        _searchQuery.value = value
    }

    fun setCategory(category: String?) {
        _selectedCategory.value = category?.takeIf { it.isNotBlank() }
    }

    fun setSortMode(mode: EventSortMode) {
        _sortMode.value = mode
    }

    fun refresh() {
        viewModelScope.launch {
            _isLoading.value = true
            syncStateRepository.setRunning(syncEntity)
            val preferredLanguage = resolvePreferredLanguage()
            val result = runCatching { eventsRepository.getEvents(preferredLanguage) }
            result.onSuccess { list ->
                _events.value = list
                syncStateRepository.setSuccess(syncEntity)
            }.onFailure { throwable ->
                syncStateRepository.setError(syncEntity, throwable.message)
            }
            _isLoading.value = false
        }
    }

    private suspend fun resolvePreferredLanguage(): String {
        val mode = userPreferenceRepository.getString("ui.lang.mode", "auto")
        val manual = userPreferenceRepository.getString("ui.lang.manual", "uk").ifBlank { "uk" }
        val preferred = if (mode == "manual") {
            manual.lowercase()
        } else {
            Locale.getDefault().language.ifBlank { "uk" }.lowercase()
        }
        return if (preferred in supportedEventLanguages) preferred else "uk"
    }

    fun retrySyncInBackground() {
        SyncScheduler.scheduleEventsRetry(workManager)
    }

    fun selectEvent(event: EventItem?) {
        _selectedEvent.value = event
    }

    fun openSelectedEventOnMap() {
        val event = _selectedEvent.value ?: return
        val lat = event.latitude ?: return
        val lon = event.longitude ?: return

        viewModelScope.launch {
            userPreferenceRepository.upsert("map.focus.lat", lat.toString())
            userPreferenceRepository.upsert("map.focus.lon", lon.toString())
            userPreferenceRepository.upsert("map.focus.zoom", "15")
            userPreferenceRepository.deleteByKey("map.focus.placeId")
        }
    }

    fun addSelectedEventToActiveRoute() {
        val event = _selectedEvent.value ?: return
        val lat = event.latitude ?: return
        val lon = event.longitude ?: return

        viewModelScope.launch {
            val routes = routeRepository.getAllRoutes().first()
            val active = routes.firstOrNull() ?: return@launch
            routeRepository.updateRoute(
                RouteMetricsCalculator.withAppendedWaypoint(
                    route = active,
                    waypoint = RoutePoint(
                        latitude = lat,
                        longitude = lon,
                        name = event.title
                    )
                )
            )
        }
    }
}

File: iosApp/PutivnykIOS/ViewModels/EventsViewModel.swift
import SwiftUI
import PutivnykShared

@MainActor
final class EventsViewModel: ObservableObject {
    enum SortMode: String, CaseIterable { case date, title, category }

    @Published var events: [EventItem] = []
    @Published var searchQuery: String = ""
    @Published var selectedCategory: String? = nil
    @Published var sortMode: SortMode = .date
    @Published var isLoading: Bool = false
    @Published var availableCategories: [String] = []

    private let services = AppServices.shared
    private var allEvents: [EventItem] = []

    init() { refresh() }

    func refresh() {
        Task {
            isLoading = true
            let language = await resolveLanguage()
            do {
                let fetched = try await services.eventsRepository.getEvents(language: language)
                allEvents = fetched as? [EventItem] ?? []
            } catch {
                // keep existing events on failure
            }
            availableCategories = Array(Set(allEvents.map { $0.category.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty })).sorted()
            applyFilters()
            isLoading = false
        }
    }

    func updateSearchQuery(_ value: String) {
        searchQuery = value
        applyFilters()
    }

    func setCategory(_ cat: String?) {
        selectedCategory = cat
        applyFilters()
    }

    func setSortMode(_ mode: SortMode) {
        sortMode = mode
        applyFilters()
    }

    func selectEvent(_ event: EventItem?) { }

    func openEventOnMap(_ event: EventItem) {
        guard let lat = event.latitude, let lon = event.longitude else { return }
        Task {
            try? await services.userPreferenceRepository.upsert(key: "map.focus.lat", value: "\(lat)")
            try? await services.userPreferenceRepository.upsert(key: "map.focus.lon", value: "\(lon)")
            try? await services.userPreferenceRepository.upsert(key: "map.focus.zoom", value: "15")
        }
    }

    private func applyFilters() {
        var filtered = allEvents
        let query = searchQuery.lowercased().trimmingCharacters(in: .whitespaces)

        if !query.isEmpty {
            filtered = filtered.filter {
                $0.title.lowercased().contains(query) ||
                $0.description_.lowercased().contains(query) ||
                $0.category.lowercased().contains(query) ||
                $0.locationName.lowercased().contains(query)
            }
        }

        if let cat = selectedCategory, !cat.isEmpty {
            filtered = filtered.filter { $0.category.caseInsensitiveCompare(cat) == .orderedSame }
        }

        switch sortMode {
        case .date: filtered.sort { $0.startsAt > $1.startsAt }
        case .title: filtered.sort { $0.title.lowercased() < $1.title.lowercased() }
        case .category: filtered.sort { $0.category.lowercased() < $1.category.lowercased() }
        }

        events = filtered
    }

    private func resolveLanguage() async -> String {
        let mode = (try? await services.userPreferenceRepository.getString(key: "ui.lang.mode", defaultValue: "auto")) ?? "auto"
        if mode == "manual" {
            return (try? await services.userPreferenceRepository.getString(key: "ui.lang.manual", defaultValue: "uk")) ?? "uk"
        }
        return Locale.current.language.languageCode?.identifier ?? "uk"
    }
}
